### PR TITLE
✨ Check lint of scaffolded projects in CI

### PR DESCRIPTION
**Pull request by Eileen-Yu**
## Description:
Add golangci-lint check for the projects scaffolds within kubebuilder in CI.

## Motivation:
1. Ensure the project quality
2. Reduce maintenance trifles
3. Aims to solve https://github.com/kubernetes-sigs/kubebuilder/issues/3119

### CODE REVIEW

File: .github/workflows/lint.yml
Comment (camilamacedo86) [inline .github/workflows/lint.yml:36]: @Eileen-Yu,

I was looking this one and I think we have a better way to do that.

See that in our CI we use the lint https://github.com/kubernetes-sigs/kubebuilder/blob/master/.github/workflows/lint.yml#L40 for the Kubebuilder code base.

The github action is nice because is visual and create comments in the pr code to let us know what is wrong.
Then, I checked and it has an option where we can pass the directories 

See:

```
# Optional: working directory, useful for monorepos
          # working-directory: somedir
``` 

From : https://github.com/golangci/golangci-lint-action

So, I think we could:
a) Copy and pate the .github/workflows/lint.yml to .github/workflows/lint_testdata.yml
b) We probably will need to create many jobs one for each testdata project
c) We could have the make lint-testdata and make linte-fix-testdata which will be targets that will do the same that we do for Kubebuilder but for the testdatda projects with a shell script as you did . So that as a dev, we can run them locally to test things out.   

WDYT?

File: Makefile
Comment (camilamacedo86) [inline Makefile:92]:
```suggestion
.PHONY: lint-testada
lint-testada: golangci-lint ## Run golangci-lint for the lscaffolded projects under testdata directory
	./lint-testada
```
Could we call something like lint-testdata or lint-samples to be more intuitive

File: lint-testdata.sh
Comment (camilamacedo86) [inline lint-testdata.sh:20]: why do need this line?

File: lint-testdata.sh
Comment (camilamacedo86) [inline lint-testdata.sh:24]: what about we have switch here where we can call both: 
$GOLANGCI_LINT run and $GOLANGCI_LINT run --fix

File: .github/workflows/lint.yml
Comment (camilamacedo86) [inline .github/workflows/lint.yml:45]: It seems not working because we have no failures and I do not think that the projects that will scaffold would pass successfully in the check.

File: .github/workflows/lint.yml
Comment (Eileen-Yu) [reply .github/workflows/lint.yml]: Sounds good 👍

I tried to apply reusable workflow to drop duplications. 

I set the reference as the master branch, which supposed to be able to work when this PR gets merged. Would this be good?

File: lint-testdata.sh
Comment (Eileen-Yu) [reply lint-testdata.sh]: I dropped this as it is not necessary any more.
Thx for pointing out!

File: lint-testdata.sh
Comment (Eileen-Yu) [reply lint-testdata.sh:28]: This sounds better. I tried to implement in the latest update. 

Would that look good to you?

File: .github/workflows/lint-testdata.yml
Comment (Kavinjsir) [inline .github/workflows/lint-testdata.yml:10]: I guess we will be using `lint-testdata-template.yml` here?

**Review by Kavinjsir — CHANGES_REQUESTED** [review]

File: .github/workflows/lint-testdata.yml
Comment (camilamacedo86) [inline .github/workflows/lint-testdata.yml:15]: I think we can test only for go/v4 and no longer care about v3.
We are close to make the go/v4 stable and make the default scaffold which would mean
deprecate v3.

**Review by camilamacedo86 — COMMENTED** [review]
Hi @Eileen-Yu,

The implementation shows fine. Just a few nits : https://github.com/kubernetes-sigs/kubebuilder/pull/3216/files#r1105460858 and https://github.com/kubernetes-sigs/kubebuilder/pull/3216/files#r1105340810

However, when you run locally the check for the testdata is that working?
Are you seen issues? I do not think that the projects would pass on that.

File: .github/workflows/lint-testdata.yml
Comment (Eileen-Yu) [reply .github/workflows/lint-testdata.yml]: Thx for catching! 👍

**Comment by Eileen-Yu** [comment]
> However, when you run locally the check for the testdata is that working? Are you seen issues? I do not think that the projects would pass on that.

Yes unfortunately there are lots of lint error :(
Would pick some time to try to fix.

**Comment by camilamacedo86** [comment]
Hi @Eileen-Yu,

> Yes unfortunately there are lots of lint error :(

So, this CI check should fail right :-) 
We need to fix that. I would expected we check the comments with the lint issues on it. 
OR maybe it will only occurs after we  get it merged for a next follow up PR? 
That might be the reason for we not seem them. 

> Because these projects are ONLY scaffolded by kubebuilder commands.
> We use them to validate the code consistency.
> If we want to enforce lint to them, then we will have to modify the content of them right?

The goal is to force otherwise we think that all is OK when it is not. 
And yes, we will need to fix the issues. 
However, we can:
- a) Just check and fix go/v4 ones (go/v3 will be deprecated soon)
- b) See the errors here first and them open another PRs to sort them out.

File: .github/workflows/lint-testdata-template.yml
Comment (camilamacedo86) [inline .github/workflows/lint-testdata-template.yml:31]:
```suggestion
```

Comment (camilamacedo86) [reply .github/workflows/lint-testdata-template.yml:31]: I think is because of it we are not checking the issue right?

Comment (camilamacedo86) [reply .github/workflows/lint-testdata-template.yml:31]: However, I am OK with either we move in this way and then create a task for we sort out all issues as follow up.
What do you think ?

File: .github/workflows/lint-testdata.yml
Comment (everettraven) [inline .github/workflows/lint-testdata.yml:10]: So I think using a template is a really neat approach, but not sure it is necessary. Since we already have a target being added to the Makefile that *should* automatically capture any new testdata directories I think it would be easier to maintain an action workflow that simply calls that make target. 

That being said, if there is agreement from other maintainers on taking this approach I am fine with it as well.

**Comment by ashutosh887** [comment]
@Eileen-Yu let me know if you need any help resolving the Issue. Let's resolve this together if you want 😊✨

File: Makefile
Comment (camilamacedo86) [reply Makefile:96]: I do not think so, we should not run fix-lint in the CI
It is ONLY a helper for us in the develop env and does not fix all issues automatically.
It can only fix issues like "go fmt" and extra imports.

Comment (everettraven) [reply Makefile:96]: >It is a helper for dev env and not be called in the CI.
We should not try to fix issues in the CI at all.
Also, it does not fix all issues automatically.
It can only fix issues like "go fmt" and extra imports.
>
>I mean:
>
  >  In the CI we just check
  >  Locally, in dev env, we can use make lint-fix to help out.
>
>It is for we have the same make lint-fix that we have for the KB code to the samples either.

Yeah, but all the projects under the `testdata/` directory are automatically generated. Using a lint fixing command will fix the problem but the problem will come back when the testdata is regenerated if it isn't fixed in the actual scaffold logic.

**Comment by Sajiyah-Salat** [comment]
@camilamacedo86  whats the reason that this pr is still just open without any activity.

**Comment by camilamacedo86** [comment]
I have been thinking on this one.
We could either consider create a new plugin like QA plugin which would:
- Add the lint target 
- Add GitHub actions 
- Add tests and etc

I mean, todo scaffolds on top to ensure the quality of the projects following the common practices adopt by the community. 

WDYT 

@Kavinjsir @Eileen-Yu ?

**Comment by Kavinjsir** [comment]
> I have been thinking on this one. We could either consider create a new plugin like QA plugin which would:
> 
> * Add the lint target
> * Add GitHub actions
> * Add tests and etc
> 
> I mean, todo scaffolds on top to ensure the quality of the projects following the common practices adopt by the community.
> 
> WDYT
> 
> @Kavinjsir @Eileen-Yu ?

@camilamacedo86 I think introducing a new plugin to help users with QA is a good idea.

From my view of the previous context, it seems challenging for the Kubebuilder repository `itself` to deal with lint check on the 'users code'. (The scaffolded code is never complete.)

For your suggestions on the proposed QA plugin, I think:
1. It is good to introduce `golangci-lint` for code quality since it looks being used widely.
2. I'm not sure if it is necessary to provide `github action` template also, since there are also other common CI/CD tools/platforms.
3. I'm not sure quite follow the idea to `add test`, are you suggesting making scaffold to code lines for unit tests OR e2e tests?

**Comment by varshaprasad96** [comment]
This was brought up in the community meeting, and one of the suggestion could be to scaffold out as a `Makefile` command (as in here: https://github.com/operator-framework/helm-operator-plugins/blob/2bd2a0ab53eed258af11beee1cf8c2537a3995dc/Makefile#L64) instead of introducing a new command or plugin for this. This would make it easier for the plugin author too, to just run `make lint` instead of using a new command altogether (having a separate command for checking link might be an overkill imo).

**Pull request closed without merging**
